### PR TITLE
Feature/keyboard enhancements

### LINKED ISSUE #181: [SDL 0238] Keyboard Enhancements
Proposal: [Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0238-Keyboard-Enhancements.md)

> This proposal is about adding a new keyboard layout and adding new properties for keyboards supported by SDL. The following keyboard enhancements are discussed in this proposal.

Enhancement #1: Support for Numeric Keyboard.

Enhancement #2: Allow apps to mask entered characters.

Enhancement #3: Allow apps to change special characters shown on the keyboard layout.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/754

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the revisions outlined in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/754#issuecomment-511379464) from the author.  The project maintainer also agreed to the author's point regarding `INVALID_DATA` (12).

> The proposal .md file was updated to reflect these revisions on 7/22/19.

### PR DESCRIPTION
**Pull request by JackLivio**
Fixes #181 
### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Core Tests
- Perform interaction keyboard with all layouts (numeric is new layout)
- Test behavior of masking global property
- Test behavior of setting custom keys on different layouts

Core version / branch / commit hash / module tested against:  Keyboard enhancements branch / 4e7603e604ec578c2aa5fa41784dce6ea5c60fb9
Proxy+Test App name / version / branch / commit hash / module tested against: Test suite master

### Summary
Adds support for a new layout: NUMERIC
Adds ability to mask the users input
Adds ability for mobile to set custom keys in place of special characters on the top level of the keyboard layout.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/js/Keyboard.js [hunk 1]
```diff
@@ -67,25 +68,77 @@ export default class Keyboard extends Component {
...
+    })
+    const event = toggleMask ? "INPUT_KEY_MASK_ENABLED" : "INPUT_KEY_MASK_DISABLED"
+    uiController.onKeyboardInput(null, event);
+  }
+
+  replaceSpecialCharacters(keyboardLayout, characters) {
+    var layout = keyboardLayout.default;
+    var customCharacters = characters;
+    //Special characters on root level of supported keyboards
+    var replaceableChars = '`-=[]\\;\',./^+#<&(_)=$:!@';
```
Comment (jacobkeeler) [inline src/js/Keyboard.js:84]:
```suggestion
    var replaceableChars = '`-=[]\\;\',./^+#<&(_)$:!@';
```
Seems that `=` was included twice

File: src/js/Keyboard.js [hunk 2]
```diff
@@ -98,11 +151,28 @@ export default class Keyboard extends Component {
...
+                    onClick={this.handleUserMask}
+                  />
+                  <label 
+                    for="maskOption" 
+                    className="mask-checkbox mask-option-label"
+                    style={{display: showUserMaskOption ? 'inline' : 'none' }}
+                  >
+                    Mask Input
+                  </label>
+                </div>
```
Comment (jacobkeeler) [inline src/js/Keyboard.js:175]:
```suggestion
                <div className="input-row">
                    <input
                        className="input-text"
                        value={this.state.input}
                        type={maskedInput ? "password" : "text"}
                        placeholder={"Tap on the virtual keyboard to start"}
                        onChange={this.onChangeInput}
                    />
                    <input 
                        className="mask-checkbox"
                        id="maskOption"
                        type={showUserMaskOption ? "checkbox" : "hidden"} 
                        onClick={this.handleUserMask}
                    />
                    <label 
                        for="maskOption" 
                        className="mask-checkbox mask-option-label"
                        style={{display: showUserMaskOption ? 'inline' : 'none' }}
                    >
                        Mask Input
                    </label>
                </div>
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

**Review by jacobkeeler — CHANGES_REQUESTED** [review]
Also, it looks like keyboard property changes do not take effect while the keyboard is active unless a button is pressed, which can be a bit jarring. I think that ideally the keyboard should update immediately if the keyboard properties are set while it is active.

File: src/js/Controllers/UIController.js [hunk 3]
```diff
@@ -236,10 +236,25 @@ class UIController {
...
                 ValidateImages([rpc.params.menuIcon]).then(
-                    () => {this.listener.respondSuccess(rpc.method, rpc.id)},
-                    () => {this.listener.send(RpcFactory.InvalidImageResponse(rpc))}
+                    () => {
+                        if (warningsString) {
+                            this.listener.send(RpcFactory.ErrorResponse(rpc, 21, warningsString))
+                        } else {
+                            this.listener.respondSuccess(rpc.method, rpc.id)
+                        }
+
```
Comment (jacobkeeler) [inline src/js/Controllers/UIController.js:253]: Nitpick

**Review by jacobkeeler — APPROVED** [review]

**Comment by JackLivio** [comment]
For anyone wondering, we decided that a SetGlobalProperties should not change the keyboard layout while in view.

**Pull request merged**